(PDB-2242) Move from moustache to comidi

senior opened the pull request:
This series of commits moves PuppetDB from moustache to comidi and refactors routes and handlers. All query routes are centralized in a single namespace, parameterized by version.

More refactoring work needs to be done. Specifically there's some rough edges around adding the parent check to routes and how URL decoding route segments is done.

Upgrading comidi to the latest (0.3.1) also required updating some other depdencies which is included in this PR as well.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/390/
Test FAILed.

kbarber commented on test/puppetlabs/puppetdb/metrics/core_test.clj:
```diff
@@ -72,6 +72,7 @@
                     :let [response (app
                                     (get-request
                                      (str version uri)))]]
+              (println                        (str version uri))
```
debugging

kbarber replied:
sorry this is wip isn't it, carry on :-).

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/393/
Test PASSed.

ajroetker commented on src/puppetlabs/puppetdb/http/command.clj:
```diff
@@ -6,12 +6,12 @@
-            [net.cgrand.moustache :as moustache]
             [puppetlabs.puppetdb.config :as conf]
             [puppetlabs.puppetdb.middleware :as mid]
             [compojure.core :as compojure]
```
This isn't used anymore

ajroetker commented on src/puppetlabs/puppetdb/meta.clj:
```diff
@@ -1,16 +1,18 @@
-             :refer [verify-accepts-json wrap-with-globals validate-no-query-params]]
+            [puppetlabs.puppetdb.middleware :as mid]
             [puppetlabs.trapperkeeper.core :refer [defservice]]
             [compojure.core :as compojure]
```
This isn't used anymore

kbarber replied:
Indeed.

ajroetker commented on project.clj:
```diff
@@ -65,15 +65,14 @@
                  ;; WebAPI support libraries.
-                 [net.cgrand/moustache "1.1.0" :exclusions [ring/ring-core org.clojure/clojure]]
-                 [compojure "1.3.3"]
+                 [compojure "1.4.0"]
```
Can we remove this dependency now?

senior replied:
Not yet. I forgot that we still use some stuff from compojure. pdb-routing.clj is a good example. I don't see anything we're using compojure for that wouldn't be pretty easy to move over to comidi. Once we've done that, we'll have no code that references compojure and I'll feel comfortable removing it.

ajroetker replied:
Should we ticket that work then for another patch?

kbarber replied:
I think another patch/ticket is a good idea, we definitely want to close the book on this soon though.

ajroetker commented on src/puppetlabs/puppetdb/http/handlers.clj:
```diff
@@ -0,0 +1,385 @@
+   
+   (cmdi/ANY ["/" :hash "/logs"] []
+             (-> (report-data-responder version "report_logs")
+                 (comp decode-route-params)
```
seems like decode-route-params should just be part of `wrap-with-puppetdb-middleware`, i.e. a global middleware

senior replied:
I wasn't confident that the middleware would be applied at the right time, I can try it out though.

ajroetker commented on src/puppetlabs/puppetdb/http/handlers.clj:
```diff
@@ -0,0 +1,385 @@
+   handler-to-append :- handler-schema]
+  (cmdi/wrap-routes route #(comp % handler-to-append)))
+
+(pls/defn-validated create-handler :- handler-schema
```
could we call this create-query-handler, or even just query-handler?

ajroetker commented on src/puppetlabs/puppetdb/middleware.clj:
```diff
@@ -342,18 +344,37 @@
+          (when handler
+            (bring/request
+             (handler-fn handler)
+             (update-in req [:route-params] merge route-params)
```
`(update req :route-params...`

ajroetker commented:
Lgtm for the most part but FYI this causes PE to not be able to compile, otherwise :+1:

senior commented:
@ajroetker I have a PR up for the PE side, it'll need to be updated once I push the change of create-handler -> create-query handler, but works as-is now. I'll investigate the decode params thing in the morning.

senior commented:
@ajroetker Addressed your comments. Renamed `create-handler` to `create-query-handler` and made the related updates. URL decoding route-params in a middleware didn't work but I did modify make-pdb-handler to do the right thing and decode it there. Looking at that function now, seems like the right place for it anyway.

I made a separate commit with the changes so it should be easy to review on it's own.

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/410/
Test PASSed.

kbarber commented:
@senior let me do some poking at this now.

kbarber commented on src/puppetlabs/puppetdb/http/v4.clj:
```diff
@@ -1,75 +1,24 @@
-
-   ["reports" &]
-   {:any (reports/reports-app version)}))
+  (apply cmdi/routes
```
Why not collapse this file completely, into server.clj? Then we can see v5 routes next to v4 routes next to each other :-).

ajroetker replied:
:+1:  to that, I never liked the namespace per api version convention

kbarber commented on test/puppetlabs/puppetdb/acceptance/cli.clj:
```diff
@@ -15,7 +15,7 @@
 
 (deftest test-anonymized-export-roundtrip
-  (doseq [profile (keys anon/anon-profiles)]
+  (doseq [profile  (take 1 (keys anon/anon-profiles))]
```
You'll be incurring whitespace tax here, lol

senior replied:
No idea why this change was made, I'll roll it back

kbarber commented:
@senior 3 things:

* That cli.clj thingy
* Consider collapsing v4 -> server.clj
* The left over reference to compojure that AJ mentioned

Otherwise I'm +1, the new layout feels good to me, compressing this stuff into handlers.clj means the problem is together, and we can start to apply code reduction easier now. Well done.

senior commented:
@kbarber @ajroetker v4 is now collapsed into server.clj, that stray compojure reference has been removed, same with the inadvertant cli test change

puppetlabs-jenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins-enterprise.delivery.puppetlabs.net/job/enterprise_puppetdb_init-multijob_githubpr-master/414/
Test PASSed.
